Kernel/Process.cpp

```diff
@@ -790,21 +785,39 @@ int Process::sys$restore_signal_mask(u32 mask)
     return 0;
 }
 
-void Process::sys$sigreturn()
-{
-    InterruptDisabler disabler;
-    Scheduler::prepare_to_modify_tss(*current);
-    current->m_tss = *current->m_tss_to_resume_kernel;
-    current->m_tss_to_resume_kernel.clear();
-#ifdef SIGNAL_DEBUG
-    kprintf("sys$sigreturn in %s(%u)\n", name().characters(), pid());
-    auto& tss = current->tss();
-    kprintf(" -> resuming execution at %w:%x stack %w:%x flags %x cr3 %x\n", tss.cs, tss.eip, tss.ss, tss.esp, tss.eflags, tss.cr3);
-#endif
-    current->set_state(Thread::State::Skip1SchedulerPass);
-    Scheduler::yield();
-    kprintf("sys$sigreturn failed in %s(%u)\n", name().characters(), pid());
-    ASSERT_NOT_REACHED();
+int Process::sys$sigreturn(RegisterDump& registers)
+{
+    u32* stack_ptr = (u32*)registers.esp_if_crossRing;
+    u32 smuggled_eax = *stack_ptr;
+    stack_ptr++; //the stored eax
+    stack_ptr++; //ebp
+    stack_ptr++; //return adress
+    stack_ptr++; //handler
+    stack_ptr++; //signal
+    stack_ptr++; //old signal mask
+    registers.edi = *stack_ptr;
```

Comment:
Things were arranged in this specific order because I was doing a fake `PUSHA` in C++ and then doing a literal `POPA` in assembly.
Now were are doing both fake `PUSHA` and `POPA` so it's not really necessary to adhere to this specific structure. Instead we could just `memcpy` in and out of a `RegisterDump` struct :)